src/span.h

```diff
@@ -4,40 +4,90 @@
 #ifndef USE_CXX17
 #include <span>
 #else
+
 #include <array>
+#include <stdexcept>
 #include <vector>
-namespace std {
 
+namespace std {
+namespace generators_span {
 template <typename T>
 struct span {
   span() = default;
-  span(T* p, size_t length) : p_{p}, length_{length} {}
+  constexpr span(T* p, size_t length) noexcept : p_{p}, length_{length} {}
+  constexpr span(const span<T>& s) noexcept : p_{s.p_}, length_{s.length_} {}
 
-  span(const span<std::remove_const_t<T> >& s) : p_{const_cast<T*>(s.data())}, length_{s.size()} {}
-  span(const std::vector<std::remove_const_t<T> >& s) : p_{const_cast<T*>(s.data())}, length_{s.size()} {}
+  span(std::vector<T>& s) noexcept : p_{s.data()}, length_{s.size()} {}
   template <size_t N>
-  span(const std::array<std::remove_const_t<T>, N>& s) : p_{const_cast<T*>(s.data())}, length_{s.size()} {}
+  constexpr span(std::array<T, N>& s) noexcept : p_{s.data()}, length_{s.size()} {}
 
-  bool empty() const { return length_ == 0; }
+  constexpr bool empty() const noexcept { return length_ == 0; }
 
-  size_t size() const { return length_; }
-  size_t size_bytes() const { return length_ * sizeof(T); }
+  constexpr size_t size() const noexcept { return length_; }
+  constexpr size_t size_bytes() const noexcept { return length_ * sizeof(T); }
 
-  T* data() { return p_; }
-  const T* data() const { return p_; }
+  constexpr T* data() const noexcept { return p_; }
 
-  T& operator[](size_t index) const { return p_[index]; }
+  constexpr T& operator[](size_t index) const noexcept { return p_[index]; }
 
-  T& back() const { return p_[length_ - 1]; }
+  constexpr T& back() const noexcept { return p_[length_ - 1]; }
 
-  T* begin() const { return p_; }
-  T* end() const { return p_ + length_; }
+  constexpr T* begin() const noexcept { return p_; }
+  constexpr T* end() const noexcept { return p_ + length_; }
 
-  span subspan(size_t index, size_t length) const { return span(p_ + index, length); }
+  span subspan(size_t index, size_t length) const {
```

Comment:
https://en.cppreference.com/w/cpp/container/span/subspan
some differences:
- `subspan(size(), 0)` should also be valid.
- there's a default argument value for `length` of `std::dynamic_extent` which has special handling.
- undefined behavior vs. exception.

not sure if the second two are worth implementing now, but maybe the index/length validation can be updated.

"The behavior is undefined if either Offset or Count is out of range. This happens if
- Offset is greater than size(), or
- Count is not std::dynamic_extent and Count is greater than size() - Offset."

Reply:
I intentionally veered off the undefined behavior. I will update the index validation.

Reply:
Made subspan(size(), 0) valid. Kept exceptions. It would not harm compatibility.